src/sharedring.rs

```diff
@@ -29,34 +30,80 @@ fn round_to_page_size<T>(capacity: usize) -> usize {
     let bytes = crate::ringbuf::channel_bufsize::<T>(capacity);
     let ps = page_size();
     let m = bytes % ps;
-    if m == 0 { bytes } else { bytes + ps - m }
+    if m == 0 {
+        bytes
+    } else {
+        bytes + ps - m
+    }
 }
 
 fn eventfd() -> Result<File, std::io::Error> {
     let x = unsafe { libc::eventfd(0, libc::EFD_CLOEXEC) };
-    if x == -1 { Err(std::io::Error::last_os_error()) }
-    else { Ok(unsafe { File::from_raw_fd(x) }) }
+    if x == -1 {
+        Err(std::io::Error::last_os_error())
+    } else {
+        Ok(unsafe { File::from_raw_fd(x) })
+    }
 }
 
 impl Inner {
     fn new<T>(capacity: usize) -> Result<Self, Error> {
+        Self::with_opts::<T>(capacity, None)
+    }
+
+    fn with_hugetlb<T>(capacity: usize, tlbsize: HugetlbSize) -> Result<Self, Error> {
+        Self::with_opts::<T>(capacity, Some(tlbsize))
+    }
+
+    fn with_opts<T>(capacity: usize, tlbsize: Option<HugetlbSize>) -> Result<Self, Error> {
         let bytes = round_to_page_size::<T>(capacity);
-        let opts = crate::mem::mfd::MemfdOptions::default().allow_sealing(true).close_on_exec(true);
+        let mut opts = MemfdOptions::default()
+            .allow_sealing(true)
+            .close_on_exec(true);
+        if tlbsize.is_some() {
+            opts = opts.hugetlb(tlbsize);
+        }
+
         let memfd = opts.create(std::any::type_name::<T>())?;
-        memfd.as_file().set_len(bytes as u64)?;
+        if tlbsize.is_none() {
+            // hugetlb does not need/allow to set_len
+            memfd.as_file().set_len(bytes as u64)?;
```

Comment:
Here's something weird. If the file's length is given by the hugetlbsize, nothing is stopping us from setting a large capacity and then a smaller hugetlbsize. This looks like it at least needs to be checked for somehow, or possibly automatically calculated?

Reply:
hugetlbsize is the size of a special page called huge-page. Usually, a CPU provides 2MB as the option, and some CPUs provide 1GB or others. Say, if bytes is 8MB, then this memfd will be backed by 4 2MB huge-page. hugetlbsize is not the size of the memfd.